### PR TITLE
(+) #333 CTL-271 Support generic lists in ServiceLocator

### LINKED ISSUE #333: CTL-271: Support generic lists in ServiceLocator
Jira issue originally created by user @geertvanhorrik:

When several registrations of `IService` exist in a container, many frameworks automatically provide `IService[]` (or `List<IService>` etc).

This is extremely important for open/closed principle. For example,
UserValidator can have dependency on `IUserValidationRule[]`. Then new rules can be transparently added to the system without any changes to UserValidator.

### PR DESCRIPTION
**Pull request by alexfdezsauco**
Fixes #333 .

### Checklist

- [X] I have included examples or tests
- [X] I have updated the change log

### Changes proposed in this pull request:

- Support generic lists in ServiceLocator

### CODE REVIEW

File: src/Catel.Core/Catel.Core.Shared/IoC/Interfaces/IServiceLocator.cs [hunk 1]
```diff
@@ -115,6 +115,15 @@ public interface IServiceLocator : IServiceProvider, IDisposable
...
         bool IsTypeRegisteredAsSingleton(Type serviceType, object tag = null);
 
+        /// <summary>
+        /// Determines whether the specified service type is registered tag regardless.
+        /// </summary>
+        /// <param name="serviceType">The type of the service.</param>
+        /// <returns><c>true</c> if the specified service type is registered; otherwise, <c>false</c>.</returns>
+        /// <remarks>Note that the actual implementation lays in the hands of the IoC technique being used.</remarks>
+        /// <exception cref="ArgumentNullException">The <paramref name="serviceType"/> is <c>null</c>.</exception>
+        bool IsTypeRegisteredTagRegardless(Type serviceType);
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/IoC/Interfaces/IServiceLocator.cs]: In what context is "regardless" meant? I'm not sure whether it's the right name for this method.

File: src/Catel.Core/Catel.Core.Shared/IoC/ServiceLocator.cs [hunk 2]
```diff
@@ -656,6 +658,74 @@ public void RemoveAllTypes(Type serviceType)
...
         #region Methods
+        /// <summary>
+        /// Determines whether the specified service type is registered as open generic.
+        /// </summary>
+        /// <param name="serviceType">The type of the service.</param>
+        /// <param name="tag">The tag to register the service with. The default value is <c>null</c>.</param>
+        /// <returns><c>true</c> if the specified service type is registered; otherwise, <c>false</c>.</returns>
+        private bool IsTypeRegisteredAsOpenGeneric(Type serviceType, object tag = null)
+        {
+            if (serviceType.IsGenericTypeEx())
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/IoC/ServiceLocator.cs]: Now this has been migrated to a new method, we can simplify it (if (!serviceType.IsGenerictypeEx) return

File: src/Catel.Core/Catel.Core.Shared/IoC/TypeFactory.cs [hunk 3]
```diff
@@ -416,8 +417,17 @@ private bool CanConstructorBeUsed(ConstructorInfo constructor, object tag, bool
                         var parameterToResolve = ctorParameters[j];
                         var parameterTypeToResolve = parameterToResolve.ParameterType;
 
-                        if (!_serviceLocator.IsTypeRegistered(parameterTypeToResolve, tag) && 
-                            !_serviceLocator.IsTypeRegistered(parameterTypeToResolve))
+                        if (!typeof(string).IsAssignableFrom(parameterTypeToResolve) && typeof(IEnumerable).IsAssignableFrom(parameterTypeToResolve))
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/IoC/TypeFactory.cs]: IsAssignableFromEx (keep xplat as much as possible)

File: src/Catel.Core/Catel.Core.Shared/IoC/TypeFactory.cs [hunk 4]
```diff
@@ -416,8 +417,17 @@ private bool CanConstructorBeUsed(ConstructorInfo constructor, object tag, bool
...
+                        {
+                            var collectionElementType = parameterTypeToResolve.GetCollectionElementType();
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/IoC/TypeFactory.cs]: I think we have an Ex implementation for this as well.

File: src/Catel.Core/Catel.Core.Shared/Linq/Extensions/EnumerableExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="EnumerableExtensions.cs" company="Catel development team">
+//   Copyright (c) 2008 - 2017 Catel development team. All rights reserved.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+
+namespace Catel.Linq.Extensions
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/Linq/Extensions/EnumerableExtensions.cs]: Remove .Extensions

File: src/Catel.Core/Catel.Core.Shared/Linq/Extensions/EnumerableExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    using Catel.Collections;
+    using Catel.Reflection;
+
+    /// <summary>
+    /// The enumerable extensions class.
+    /// </summary>
+    public static class EnumerableExtensions
+    {
+        /// <summary>
+        /// The cast method info.
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/Linq/Extensions/EnumerableExtensions.cs]: No need to xml document private members (especially fields).

File: src/Catel.Core/Catel.Core.Shared/Reflection/Extensions/TypeExtensions.cs [hunk 7]
```diff
@@ -173,21 +173,27 @@ public static Type GetCollectionElementType(this Type type)
...
+                return type.GetElementTypeEx();
             }
 
-            var interfaces = type.GetInterfacesEx();
-            var genericEnumerableInterface = (from iface in interfaces
-                                              where iface.Name.StartsWith("IEnumerable") && iface.IsGenericTypeEx()
-                                              select iface).FirstOrDefault();
-            if (genericEnumerableInterface == null)
+            Type genericEnumerableInterface;
+            if (typeof(IEnumerable).IsAssignableFromEx(type) && type.IsGenericType)
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/Reflection/Extensions/TypeExtensions.cs]: IsGenericTypeEx

File: src/Catel.Core/Catel.Core.Shared/Reflection/Extensions/TypeExtensions.cs [hunk 8]
```diff
@@ -173,21 +173,27 @@ public static Type GetCollectionElementType(this Type type)
...
+                    select iface).FirstOrDefault();
+                if (genericEnumerableInterface == null)
+                {
+                    return null;
+                }
             }
 
-            var elementType = genericEnumerableInterface.GetGenericArgumentsEx()[0];
-            return elementType;
+            return genericEnumerableInterface.GetGenericArgumentsEx()[0];
```
Comment (GeertvanHorrik) [inline src/Catel.Core/Catel.Core.Shared/Reflection/Extensions/TypeExtensions.cs]: .FirstOrDefault to be sure?

**Comment by alexfdezsauco** [comment]
Fixed

**Pull request merged**